File: ebkar_reports/ebkar_reports/report/sales_invoice_delivery_status/sales_invoice_delivery_status.js
frappe.query_reports["Sales Invoice Delivery Status"] = {
    "filters": [
        {
            "fieldname": "Customer",
            "label": __("Customer"),
            "fieldtype": "Link",
            "options": "Customer",
            "default": "",
            "reqd": 0
        },
        {
            "fieldname": "start_date",
            "label": __("Start Date"),
            "fieldtype": "Date",
            "reqd": 1
        },
        {
            "fieldname": "end_date",
            "label": __("End Date"),
            "fieldtype": "Date",
            "reqd": 1
        }
    ],
	"formatter": function(value, row, column, data) {
        let iconHTML = '';
        let color = 'black'; // اللون الافتراضي

        switch (column.fieldname) {
            case "invoice_number":
                iconHTML = `<i class="fa fa-file-invoice" style="color: blue;"></i>`; // أيقونة فاتورة
                break;
            case "customer_name":
                iconHTML = `<i class="fa fa-user" style="color: green;"></i>`; // أيقونة مستخدم
                break;
            case "total_amount":
                // إضافة الأيقونة بناءً على القيمة
                iconHTML = value < 0 
                    ? `<i class="fa fa-arrow-down" style="color: red;"></i>` 
                    : `<i class="fa fa-arrow-up" style="color: green;"></i>`;
                break;
            case "delivery_status":
                // إضافة أيقونات حالة التسليم
                if (value === "Fully Delivered") {
                    iconHTML = `<i class="fa fa-check-circle" style="color: green;"></i> Fully Delivered`;
                } else if (value === "Partially Delivered") {
                    iconHTML = `<i class="fa fa-exclamation-circle" style="color: orange;"></i> Partially Delivered`;
                } else {
                    iconHTML = `<i class="fa fa-times-circle" style="color: red;"></i> Not Delivered`;
                }
                break;
            default:
                return value; // إرجاع القيمة الأصلية إذا لم يتم العثور على الحالة
        }

        // إضافة اللون بجانب الأيقونة
        return `<span style="color: ${color};">${iconHTML} ${value}</span>`;
    },

    onload: function(report) {
        report.fields_dict['delivery_status'].formatter = function(value, row, column, data, default_formatter) {
            let iconHTML = '';
            let color = 'black'; // اللون الافتراضي

            // تخصيص الأيقونات بناءً على القيمة
            switch (column.fieldname) {
                case "delivery_status":
                    if (value === "Fully Delivered") {
                        iconHTML = `<i class="fa fa-check-circle" style="color: green;"></i> Fully Delivered`;
                    } else if (value === "Partially Delivered") {
                        iconHTML = `<i class="fa fa-exclamation-circle" style="color: orange;"></i> Partially Delivered`;
                    } else {
                        iconHTML = `<i class="fa fa-times-circle" style="color: red;"></i> Not Delivered`;
                    }
                    break;
                case "total_amount":
                    iconHTML = value < 0 ? `<i class="fa fa-arrow-down" style="color: red;"></i>` : `<i class="fa fa-arrow-up" style="color: green;"></i>`;
                    break;
                default:
                    return value; // إرجاع القيمة الأصلية إذا لم يتم العثور على الحالة
            }

            // إضافة الأيقونة واللون المناسبين
            return `<span style="color: ${color};">${iconHTML}</span>`;
        };
    }
};

File: ebkar_reports/ebkar_reports/report/sales_invoice_delivery_status/sales_invoice_delivery_status.py
import frappe

def get_delivery_status(invoice):
    # جلب عناصر Delivery Note المرتبطة بالفاتورة
    delivery_note_items = frappe.get_all('Delivery Note Item', filters={'against_sales_invoice': invoice}, fields=['parent', 'docstatus'])
    
    # التحقق من حالة التسليم بناءً على `docstatus`
    if not delivery_note_items:
        return "Not Delivered"
    
    fully_delivered = all(item.docstatus == 1 for item in delivery_note_items)
    if fully_delivered:
        return "Fully Delivered"
    else:
        return "Partially Delivered"

def execute(filters=None):
    columns = [
        {"fieldname": "invoice_number", "label": '<i class="fa fa-file-invoice"></i> Invoice Number', "fieldtype": "Link", "options": "Sales Invoice", "width": 150},
        {"fieldname": "customer_name", "label": '<i class="fa fa-user"></i> Customer Name', "fieldtype": "Link", "options": "Customer", "width": 200},
        {"fieldname": "posting_date", "label": '<i class="fa fa-calendar"></i> Invoice Date', "fieldtype": "Date", "width": 100},
        {"fieldname": "total_amount", "label": '<i class="fa fa-money-bill-wave"></i> Total Amount', "fieldtype": "Float", "width": 150},
        {"fieldname": "delivery_status", "label": '<i class="fa fa-truck"></i> Delivery Status', "fieldtype": "Data", "width": 150}
    ]
    
    # إنشاء شرط استعلام (conditions) بناءً على الفلاتر
    conditions = "WHERE i.docstatus = 1 AND i.posting_date BETWEEN %(start_date)s AND %(end_date)s"
    if filters.get('Customer'):
        conditions += " AND i.customer = %(Customer)s"
    
    # استعلام SQL لجلب بيانات الفواتير
    data = frappe.db.sql("""
        SELECT
            i.name AS invoice_number,
            i.customer AS customer_name,
            i.posting_date,
            i.grand_total AS total_amount
        FROM
            `tabSales Invoice` i
        {conditions}
    """.format(conditions=conditions), filters, as_dict=True)
    
    # تحديد حالة التسليم لكل فاتورة
    for row in data:
        row['delivery_status'] = get_delivery_status(row['invoice_number'])
    
    return columns, data